src/utils.py

```diff
@@ -0,0 +1,646 @@
+import ast
+import itertools
+import os
+from datetime import datetime
+from pathlib import Path
+from typing import Literal
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import pycountry
+import rioxarray as rxr
+import shapely
+import xarray as xr
+from ochanticipy import CodAB, create_country_config
+from rasterio.enums import Resampling
+from shapely.validation import make_valid
+from tqdm import tqdm
+
+from src.constants import asap1_seasonalcalendar2gaul
+
+DATA_DIR = Path(os.getenv("AA_DATA_DIR"))
+
+
+def process_asap_phenology_dekads(end: Literal["e", "sen"]):
+    """Takes ASAP phenology TIFs (start and end of seasons 1 and 2), and
+    outputs boolean for whether in season or not, by 1km pixel.
+    Returns
+    -------
+
+    """
+    # load ASAP TIFs
+    load_dir = DATA_DIR / "public/raw/glb/asap/reference_data"
+    das = []
+    for start_end in ["s", end]:
+        da_ses = []
+        for season in [1, 2]:
+            filestem = f"pheno{start_end}{season}_v03"
+            filename = f"{filestem}.tif"
+            da = rxr.open_rasterio(load_dir / filename)
+            da["start_end"] = start_end
+            da["season"] = season
+            da = da.squeeze(dim="band", drop=True)
+            da_ses.append(da)
+        da_se = xr.concat(da_ses, dim="season")
+        das.append(da_se)
+    da = xr.concat(das, dim="start_end")
+
+    # iterate over longitudes
+    # Note: this is probably a pretty slow way of doing this, but doing the
+    # whole thing at once takes up too much memory and crashes.
+    # It seems like even iterating over start/end and season 1/2 doesn't work.
+    # Hence the for loop arbitrarily over longitude.
+    dekads = range(1, 37)
+    da_ds = []
+    for lon in tqdm(range(-180, 181)):
+        da_f = da.sel(x=slice(lon, (lon + 1)))
+        # ensure that the order of the dims is correct
+        dims = [x for x in da_f.dims if x in ["x", "y"]] + ["dekad"]
+        da_d = xr.DataArray(
+            dims=dims, coords={"y": da_f.y, "x": da_f.x, "dekad": dekads}
+        )
+        da_d.rio.write_crs(4326, inplace=True)
+        da_d = da_d.rio.set_spatial_dims("x", "y", inplace=True)
+        # da_d = da_d.astype("uint8")
+        for dekad in dekads:
+            da_d.loc[{"dekad": dekad}] = (
```

Comment:
Not sure if more efficient, but might be able to use the modulo operator I referenced in the other PR to reduce this to just the 2 sections connected by `|` for season 1 and 2.